src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectSystem.cs

```diff
@@ -56,50 +48,55 @@ public LanguageServerProjectSystem(
-            var buildHost = await buildHostProcessManager.GetBuildHostAsync(BuildHostProcessKind.NetCore, CancellationToken.None);
-
-            // If we don't have a .NET Core SDK on this machine at all, try .NET Framework
-            if (!await buildHost.HasUsableMSBuildAsync(solutionFilePath, CancellationToken.None))
-            {
-                var kind = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? BuildHostProcessKind.NetFramework : BuildHostProcessKind.Mono;
-                buildHost = await buildHostProcessManager.GetBuildHostAsync(kind, CancellationToken.None);
-            }
+        _logger.LogInformation(string.Format(LanguageServerResources.Loading_0, solutionFilePath));
+        ProjectFactory.SolutionPath = solutionFilePath;
 
-            foreach (var project in await buildHost.GetProjectsInSolutionAsync(solutionFilePath, CancellationToken.None))
-            {
-                ProjectsToLoadAndReload.AddWork(new ProjectToLoad(project.ProjectPath, project.ProjectGuid, ReportTelemetry: true));
-            }
+        // We'll load solutions out-of-proc, since it's possible we might be running on a runtime that doesn't have a matching SDK installed,
+        // and we don't want any MSBuild registration to set environment variables in our process that might impact child processes.
+        await using var buildHostProcessManager = new BuildHostProcessManager(globalMSBuildProperties: AdditionalProperties, loggerFactory: LoggerFactory);
+        var buildHost = await buildHostProcessManager.GetBuildHostAsync(BuildHostProcessKind.NetCore, CancellationToken.None);
 
-            // Wait for the in progress batch to complete and send a project initialized notification to the client.
-            await ProjectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
-            await ProjectInitializationHandler.SendProjectInitializationCompleteNotificationAsync();
+        // If we don't have a .NET Core SDK on this machine at all, try .NET Framework
+        if (!await buildHost.HasUsableMSBuildAsync(solutionFilePath, CancellationToken.None))
+        {
+            var kind = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? BuildHostProcessKind.NetFramework : BuildHostProcessKind.Mono;
+            buildHost = await buildHostProcessManager.GetBuildHostAsync(kind, CancellationToken.None);
         }
+
+        var projects = await buildHost.GetProjectsInSolutionAsync(solutionFilePath, CancellationToken.None);
+        // TODO: this '!' is doing a "nullable covariant" conversion of the ImmutableArray tuple elements.
+        // This doesn't introduce any null safety issue as the elements can't be modified.
+        // It's not clear if there's a simple pattern that lets us get rid of this,
+        // except perhaps by making the nullabilities exactly match all the way down the chain that this value flows from.
```

Comment:
I'm OK if we just have LoadProjectsAsync take a single project and the loop still happens here.